File: csharp/023_Merged_K_Sorted_List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeKLists(ListNode[] lists) {
        ListNode head = new ListNode(), curr = head;
        List<KeyValuePair<int, ListNode>> m = new List<KeyValuePair<int, ListNode>>();

        // add all node to map
        foreach (ListNode listHead in lists) {
            curr = listHead;
            while (curr != null) {
                m.Add(new KeyValuePair<int, ListNode>(curr.val, curr));
                curr = curr.next;
            }
        }

        // sorted map
        m.Sort(delegate(KeyValuePair<int, ListNode> x, KeyValuePair<int, ListNode> y) {
            return x.Key.CompareTo(y.Key);
        });

        // link result list
        curr = head;
        foreach(KeyValuePair<int, ListNode> node in m) {
            curr.next = node.Value;
            curr = curr.next;
            curr.next = null;
        }

        return head.next;
    }
}

File: cpp/023_Merged_K_Sorted_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> q;
        ListNode *head = new ListNode(), *curr = head;

        // first add minimum element of all list
        for (ListNode* node : lists)
            if (node) q.push({node->val, node});

        // loop until final node
        while (!q.empty()) {
            curr->next = q.top().second;
            curr = curr->next;
            q.pop();
            
            if (curr->next) {
                // add next node of the same list
                q.push({curr->next->val, curr->next});
            }
        }

        return head->next;
    }
};